x-pack/apm-server/sampling/eventstorage/rw.go

```diff
@@ -0,0 +1,116 @@
+	// is greater than configured limit.
+	ErrLimitReached = errors.New("configured storage limit reached")
+)
+
+type RW interface {
+	ReadTraceEvents(traceID string, out *modelpb.Batch) error
+	WriteTraceEvent(traceID, id string, event *modelpb.APMEvent) error
+	WriteTraceSampled(traceID string, sampled bool) error
+	IsTraceSampled(traceID string) (bool, error)
+	DeleteTraceEvent(traceID, id string) error
+	Flush() error
+}
+
+type SplitReadWriter struct {
+	eventRW, decisionRW RW
+}
+
+func (s SplitReadWriter) ReadTraceEvents(traceID string, out *modelpb.Batch) error {
+	return s.eventRW.ReadTraceEvents(traceID, out)
+}
+
+func (s SplitReadWriter) WriteTraceEvent(traceID, id string, event *modelpb.APMEvent) error {
+	return s.eventRW.WriteTraceEvent(traceID, id, event)
+}
+
+func (s SplitReadWriter) WriteTraceSampled(traceID string, sampled bool) error {
+	return s.decisionRW.WriteTraceSampled(traceID, sampled)
+}
+
+func (s SplitReadWriter) IsTraceSampled(traceID string) (bool, error) {
+	return s.decisionRW.IsTraceSampled(traceID)
+}
+
+func (s SplitReadWriter) DeleteTraceEvent(traceID, id string) error {
+	return s.eventRW.DeleteTraceEvent(traceID, id)
+}
+
+func (s SplitReadWriter) Flush() error {
+	return nil
+}
+
+func (s SplitReadWriter) Close() error {
+	return nil
+}
+
+type storageLimitChecker interface {
+	DiskUsage() uint64
+	StorageLimit() uint64
+}
+
+type StorageLimitReadWriter struct {
+	checker storageLimitChecker
+	nextRW  RW
+}
+
+func NewStorageLimitReadWriter(checker storageLimitChecker, nextRW RW) StorageLimitReadWriter {
+	return StorageLimitReadWriter{
+		checker: checker,
+		nextRW:  nextRW,
+	}
+}
+
+func (s StorageLimitReadWriter) checkStorageLimit() error {
+	usage := s.checker.DiskUsage()
+	limit := s.checker.StorageLimit()
+	if limit != 0 && usage >= limit {
+		return fmt.Errorf("%w (current: %d, limit %d)", ErrLimitReached, usage, limit)
+	}
+	return nil
+}
+
+func (s StorageLimitReadWriter) ReadTraceEvents(traceID string, out *modelpb.Batch) error {
+	return s.nextRW.ReadTraceEvents(traceID, out)
+}
+
+func (s StorageLimitReadWriter) WriteTraceEvent(traceID, id string, event *modelpb.APMEvent) error {
+	if err := s.checkStorageLimit(); err != nil {
+		return err
+	}
+	return s.nextRW.WriteTraceEvent(traceID, id, event)
+}
+
+func (s StorageLimitReadWriter) WriteTraceSampled(traceID string, sampled bool) error {
+	if err := s.checkStorageLimit(); err != nil {
+		return err
+	}
+	return s.nextRW.WriteTraceSampled(traceID, sampled)
+}
+
+func (s StorageLimitReadWriter) IsTraceSampled(traceID string) (bool, error) {
+	return s.nextRW.IsTraceSampled(traceID)
+}
+
+func (s StorageLimitReadWriter) DeleteTraceEvent(traceID, id string) error {
+	// Technically DeleteTraceEvent writes, but it should have a net effect of reducing disk usage
+	return s.nextRW.DeleteTraceEvent(traceID, id)
+}
+
+func (s StorageLimitReadWriter) Flush() error {
+	return s.nextRW.Flush()
```

Comment:
This won't flush deletions on non-current partitions. Is that intentional, and is it OK? IIANM that means (non rotation-related) deletes may not be flushed.

Reply:
ditto, I'm more inclined to remove Flush from RW.